src/Libraries/Microsoft.Extensions.AI/ChatCompletion/AnonymousChatClient.cs

```diff
@@ -0,0 +1,242 @@
+        if (_sharedFunc is not null)
+        {
+            var updates = Channel.CreateBounded<StreamingChatCompletionUpdate>(1);
+
+#pragma warning disable CA2016 // explicitly not forwarding the cancellation token, as we need to ensure the channel is always completed
+            _ = Task.Run(async () =>
+#pragma warning restore CA2016
+            {
+                Exception? error = null;
+                try
+                {
+                    await _sharedFunc(arguments, async (chatMessages, options, cancellationToken) =>
+                    {
+                        await foreach (var update in InnerClient.CompleteStreamingAsync(chatMessages, options, cancellationToken).ConfigureAwait(false))
+                        {
+                            await updates.Writer.WriteAsync(update, cancellationToken).ConfigureAwait(false);
+                        }
+                    }).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    error = ex;
+                    throw;
+                }
+                finally
+                {
+                    _ = updates.Writer.TryComplete(error);
+                }
+            });
+
+            return updates.Reader.ReadAllAsync(cancellationToken);
+        }
+        else if (_completeStreamingFunc is not null)
+        {
+            return _completeStreamingFunc(arguments);
+        }
+        else
+        {
+            Debug.Assert(_completeFunc is not null, "Expected non-null non-streaming delegate.");
+            return CompleteStreamingAsyncViaCompleteAsync(_completeFunc!(arguments));
+
+            static async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsyncViaCompleteAsync(Task<ChatCompletion> task)
+            {
+                ChatCompletion completion = await task.ConfigureAwait(false);
+                foreach (var update in completion.ToStreamingChatCompletionUpdates())
+                {
+                    yield return update;
+                }
+            }
+        }
+    }
+
+    /// <summary>Throws an exception if both of the specified delegates are null.</summary>
+    /// <exception cref="ArgumentNullException">Both <paramref name="completeFunc"/> and <paramref name="completeStreamingFunc"/> are <see langword="null"/>.</exception>
+    internal static void ThrowIfBothDelegatesNull(
+        Func<Arguments, Task<ChatCompletion>>? completeFunc,
+        Func<Arguments, IAsyncEnumerable<StreamingChatCompletionUpdate>>? completeStreamingFunc)
+    {
+        if (completeFunc is null && completeStreamingFunc is null)
+        {
+            Throw.ArgumentNullException(nameof(completeFunc), $"At least one of the {nameof(completeFunc)} or {nameof(completeStreamingFunc)} delegates must be non-null.");
+        }
+    }
+
+    /// <summary>Represents the arguments passed to the delegates used by <see cref="AnonymousChatClient"/>.</summary>
+    public sealed class Arguments
```

Comment:
I'm guessing you chose this approach over a custom delegate to allow future extensibility/minimize lambda parameters?

Reply:
That's part of it. But also because otherwise there's a large number of parameters that must be specified in the lambda, and this consolidates them all to one. We could re-expand it if that's desirable.